kubectl explain: use openapiv3 by default

alexzielenski opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Switches kubectl explain default from buggy openapiv2 data to using new richer v3 data. The change in output should be strictly better/richer.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Switched kubectl explain to use OpenAPIV3 information published by the server. OpenAPIV2 backend can  still be used with the `--output plaintext-openapiv2` argument
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP] 3515 OpenAPI v3 for kubectl explain: https://github.com/kubernetes/enhancements/tree/master/keps/sig-cli/3515-kubectl-explain-openapiv3
```

/sig cli
/assign @apelisse

Depends on #115480 to be merged first

apelisse commented:
LGTM for the last of the 4 commits, thanks.

alexzielenski committed:
enable openapiv3 by default

alexzielenski commented:
/retest

rjsadow commented:
/release-note-edit

```release-note Switched kubectl explain to use OpenAPIV3 information published by the server. OpenAPIV2 backend can  still be used with the `--output plaintext-openapiv2` argument ```

alexzielenski commented:
/retest

alexzielenski committed:
use title case for "Object" type

for backwards compatibility

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/explain/explain.go:
```diff
@@ -82,10 +78,9 @@ type ExplainOptions struct {
 	return &ExplainOptions{
-		IOStreams:       streams,
-		CmdParent:       parent,
-		EnableOpenAPIV3: cmdutil.ExplainOpenapiV3.IsEnabled(),
```
Is it time to also remove cmdutil.ExplainOpenapiV3?

alexzielenski replied:
yes prob

ardaguclu commented:
Is this PR also fixing this issue https://github.com/kubernetes/kubernetes/issues/113902?

alexzielenski committed:
use referred type's type when it is not an object

i think we should also do this when it has no GVK? objects with no GVK cant be explained

alexzielenski committed:
log name of mising field, not path

alexzielenski committed:
throw missing field in std err

alexzielenski committed:
fix test to look for FIELD: not RESOURCE: when a field was looked up

apelisse commented:
Thanks
/approve

lavalamp commented:
/approve

apelisse commented:
/lgtm
